### ISSUE TITLE
Incorrect linking in C++ code

### ISSUE THREAD

**Issue by albert-github** [doxygen/doxygen#10989]
In the CGAL project it was noted in an example that some instances of a function were hyperlinked (correct) and that some subsequent, identical, calls were not hyperlinked.
The, reduced, code looks like:
```
/** \file */

/// a fie
void fie()
{
  copy_face_graph (boost::make_function_output_iterator([&](const std::pair<TIN_with_info::Face_handle, Mesh::Face_index>& ff)
  {
    //{
      //color_map1 = CGAL::Color(1);
    //}
    if (ff.first->info() < 0)
      color_map2 = CGAL::Color(2);
    else
    {
      color_map3 = CGAL::Color(3);
    }
  }));
}

/// The NS
namespace CGAL
{
    /// the color class
    class Color
    {
      public:
        /// constructor
        Color(int);
    };
}
```
the result is:
![image](https://github.com/doxygen/doxygen/assets/5223533/996d56a6-1f1e-4a97-9a7f-de86536ec549)

Here the `CGAL::Color` on line 12 is hyperlinked but the `CGAL::Color` on line 15 isn't.
When removing the comment signs on the lines 8 - 10 all `CGAL::Color` ar hyperlinked.

I think the interesting rules in `code.l` are here:
```
<Body,FuncCall>"{" 
```
and
```
<MemberCall2,FuncCall>"{"
```
in the later block there is also a reference to the number of open round (`(`) brackets.
Note also the double usage of the state: `FuncCall`.

As far as I can tell it is an unintended side effect from:
```
Commit: 0f01005b90986347e57a78c58585aea959da72a1 [0f01005]

Date: Friday, November 1, 2019 4:47:02 PM
```
i.e. the removal of `g_bracketCount = 0;` (now this would be `yyextra->bracketCount = 0;` in rule `<MemberCall>{SCOPETNAME}/{BN}*"("`)

@doxygen any idea what the best solution would be?

Doxygen version: 1.12.0 (e5129703e65f40fd4030d9d0c2f24561e4c3943e))
Example: [example.tar.gz](https://github.com/user-attachments/files/16120067/example.tar.gz)

**Comment by albert-github** [comment]
It looks a bit like that in the `struct ctx` of `class callContext` (in scopedtypevariant.h) besides the `name and `type` also the `bracketCount` should be stored and used for comparison in case of `theCallContext.popScope` instead a test against 0.

A bit of a spoiling part might be though that also in case of a `[` a push is used (and in case of `]` a pop.

@doxygen So I'm not sure whether this would be the right approach.

**Comment by doxygen** [comment]
@albert-github The code was not really created with lambda's in mind. But I think you are right that storing and restoring the `bracketCount` would be a good solution direction to deal with starting new bodies (i.e. `{...}`) inside a call, so that the counter can be reset to 0 at the start of a block and restored to its original value at the end of the block.

**Comment by albert-github** [comment]
Did a test with:
- LaTeX: looks OK
- Docbook: 
  ![image](https://github.com/user-attachments/assets/28537400-547f-4927-9138-330b941c8137)
  - looks like TOC is OK
  - paragraph numbering isn't. I think the `1.1.1 SECT3_md` should be numbered `1.1.0.1` (compare with LaTeX)

Note for this test I changed the `[TOC]` to `@tableofcontents{html:6,latex:6,docbook:6}`

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #11028

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.12.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
